_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+```elixir
+config :kaffe,
+  consumer: [
+    endpoints: [localhost: 9092],               
+    topics: ["our_topic", "another_topic"],     # the topic(s) that will be consumed
+    consumer_group: "example-consumer-group",   # the consumer group for tracking offsets in Kafka
+    message_handler: ExampleConsumer,           # the module that will process messages
+  ]
+```
+
+**5. Add a producer module (optional, can also call Kaffe from the console)**
+add `/lib/example_producer.ex` with the following code:
+```elixir
+defmodule ExampleProducer do
+  def send_my_message({key, value}, topic) do
+    Kaffe.Producer.produce_sync(topic, [{key, value}])
+  end
+
+  def send_my_message(key, value) do
+    Kaffe.Producer.produce_sync(key, value)
+  end
+
+  def send_my_message(value) do
+    Kaffe.Producer.produce_sync("sample_key", value)
+  end
+end
+```
+
+**6. Send and receive messages in the console!**
+Start an interactive elixir shell with `iex -S mix` and call the following:
+```sh
+iex> ExampleProducer.send_my_message({"Metamorphosis", "Franz Kafka"}, "another_topic")
+...>[debug] event#produce_list topic=another_topic
+...>[debug] event#produce_list_to_topic topic=another_topic partition=0
+...>:ok
+iex> %{
+...> attributes: 0,
+...> crc: 2125760860, # will vary
+...> key: "Metamorphosis",
+...> magic_byte: 1,
+...> offset: 1, # will vary
+...> partition: 0,
+...> topic: "another_topic",
+...> ts: 1546634470702, # will vary
+...> ts_type: :create,
+...> value: "Franz Kafka"
+...> }
+...> Metamorphosis: Franz Kafka
+```
+
+## Variations: Docker & Umbrella Apps
+- If you're running Kafka from a docker container (most common in real applications), you will use that hostname in the config file rather than `localhost`
+- In an umbrella app you'll configure Kaffe in the child application running it. If you have apps separated by environment, you can start the consumer by structuring it as a child like this:
+```elixir
+    children = case args do
+      [env: :prod] -> [worker(Kaffe.Consumer, [])]
+      [env: :test] -> []
+      [env: :dev]  -> [{ElixirKaffeCodealong.OtherApp, []}, worker(Kaffe.Consumer, [])] # note that it is not a tuple
+      [_] -> []
+    end
+```
+
+## Troubleshooting Errors
+- **No leader error**
+```
+** (MatchError) no match of right hand side value: {:error, :LeaderNotAvailable}
+```
+Solution: Try again. It just needed a minute to warm up.
+
+- **Invalid Topic error**
+```
+** (MatchError) no match of right hand side value: {:error, :InvalidTopicException}
+```
+Solution: Your topic shouldn't have spaces in it, does it?
+
+## The end
+This should have given you the basic setup for you to start exploring more of this on your own, but there's lots more you can do with Kaffe so check out sending multiple messages, consumer groups, etc. If you come up with any more troubleshooting errors you've solved, I'd love to continue adding them to what we've started here.
```

Comment:
Love the call to contribute here! Maybe we could link to https://github.com/elixirschool/elixirschool/issues or something?